zebra-chain/src/amount.rs

```diff
@@ -191,6 +191,64 @@ where
     }
 }
 
+impl<C> PartialOrd for Amount<C>
+where
+    Amount<C>: Eq,
+    C: Constraint,
+{
+    fn partial_cmp(&self, other: &Amount<C>) -> Option<std::cmp::Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl<C> Ord for Amount<C>
+where
+    Amount<C>: Eq,
+    C: Constraint,
+{
+    fn cmp(&self, other: &Amount<C>) -> std::cmp::Ordering {
+        self.0.cmp(&other.0)
+    }
+}
```

Comment:
Since we're not returning an Amount, we can compare a NonNegative amount with a NegativeAllowed amount.

We should also fix the trait bounds so they match the documented requirements in:
* https://doc.rust-lang.org/std/cmp/trait.PartialOrd.html#how-can-i-implement-partialord
* https://doc.rust-lang.org/std/cmp/trait.Ord.html#how-can-i-implement-ord

I think this constraint is required so we can call `self.cmp(other)`. But if it causes errors, just delete it:
```rust
    // We use Ord to implement PartialOrd
    Amount<C1>: Ord<Rhs = Amount<C2>>,
```

```suggestion
impl<C1, C2> PartialOrd<Rhs = Amount<C2>> for Amount<C1>
where
    Amount<C1>: PartialEq<Rhs = Amount<C2>>,
    // We use Ord to implement PartialOrd
    Amount<C1>: Ord<Rhs = Amount<C2>>,
    C1: Constraint,
    C2: Constraint,
{
    fn partial_cmp(&self, other: &Amount<C2>) -> Option<std::cmp::Ordering> {
        Some(self.cmp(other))
    }
}

impl<C1, C2> Ord<Rhs = Amount<C2>> for Amount<C1>
where
    Amount<C1>: PartialOrd<Rhs = Amount<C2>>,
    Amount<C1>: Eq<Rhs = Amount<C2>>,
    C1: Constraint,
    C2: Constraint,
{
    fn cmp(&self, other: &Amount<C2>) -> std::cmp::Ordering {
        self.0.cmp(&other.0)
    }
}
```

If the derivations don't support `PartialEq<Rhs = Amount<C2>> for Amount<C1>` and `Eq<Rhs = Amount<C2>> for Amount<C1>`, you'll need to replace them with:

```rust
impl<C1, C2> PartialEq<Rhs = Amount<C2>> for Amount<C1>
where
    // We use Ord to implement PartialEq
    Amount<C1>: Ord<Rhs = Amount<C2>>,
    C1: Constraint,
    C2: Constraint,
{
    fn eq(&self, other: &Amount<C2>) -> bool {
        match self.cmp(other) {
            std::cmp::Ordering::Equal => true,
            _ => false,
        }
    }
}

impl<C1, C2> Eq<Rhs = Amount<C2>> for Amount<C1>
where
    Amount<C1>: PartialEq<Rhs = Amount<C2>>,
    C1: Constraint,
    C2: Constraint,
{
  // Eq has no methods
}
```

Reply:
I was trying some of this and it is more complex to make it work than expected. I think is a good idea that we should implement as we will probably have use cases for it but i also think we can leave it out from this PR and open an issue to work it specifically.

Reply:
Sounds good, I added it to the tracking issue #801